### PR TITLE
[api-server] Object store folder path is configurable and can work with AWS (secure and region flag, and IAM credentials)

### PR DESCRIPTION
**Pull request by eterna2**
### Background

Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.

I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths, and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).

### Changes (Updated 16 Jan 2020)

- made `ObjectStore` folder path configurable from both config.json and env
- updated `minioclient` to use chained provider credentials: API key -> minio env flags -> aws env flags -> aws IAM
- `minioclient` `secure` and `region` flags can now also be set via `config.json` and env var.

```
	minioServiceHost       = "MINIO_SERVICE_SERVICE_HOST"
	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
	minioServiceRegion     = "MINIO_SERVICE_REGION"
	minioServiceSecure     = "MINIO_SERVICE_SECURE"
	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
	pipelineFolder         = "MINIO_PIPELINE_FOLDER"
``` 

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2080)

### CODE REVIEW

**Comment by eterna2** [comment]
ok. rebased to latest master, and fixed most of the bugs. Added a simple unit test for the new objectPath type.

File: backend/src/apiserver/client_manager.go [hunk 1]
```diff
@@ -63,6 +67,7 @@ type ClientManager struct {
 	dBStatusStore          storage.DBStatusStoreInterface
 	defaultExperimentStore storage.DefaultExperimentStoreInterface
 	objectStore            storage.ObjectStoreInterface
+	objectPaths        storage.ObjectPathsInterface
```
Comment (IronPan) [inline backend/src/apiserver/client_manager.go]: please go fmt the code.

Comment (IronPan) [inline backend/src/apiserver/client_manager.go]: creating an interface for customizing the path seems a bit overkilling. 
can we add a root path in MinioObjectStore and set the path when initializing it? When add/get file we just append root path to the filepath?

**Comment by eterna2** [comment]
hey @IronPan 
sorry for the long delay. Fixed the PR as requested.

- Added `PipelineFolder` to the config.
- Added `MINIO_PIPELINE_BUCKET_NAME` and `MINIO_PIPELINE_FOLDER` env var for consistency
- CreatePipelinePath is removed as a static func, and is now a method insider MinioObjectStore
- AddFile, DeleteFile, etc all will automatically append the base folder (PipelineFolder)
- Updated unit test to reflect these changes

**Comment by eterna2** [comment]
Updates to objectstore and corresponding unit test:

### Pipeline folder
- Added `ObjectStoreConfig.PipelineFolder` to the config.
- Added `MINIO_PIPELINE_BUCKET_NAME` and `MINIO_PIPELINE_FOLDER` env var for consistency
- `CreatePipelinePath` is removed as a static func, and is now a method insider `MinioObjectStore`
- `AddFile`, `DeleteFile`, etc all will automatically append the base folder (`PipelineFolder`)

### Fix #2819 : Region for AWS access (i.e. S3)
- Added `ObjectStoreConfig.Region` to the config.
- added `MINIO_SERVICE_REGION` env flag to set region to use for minio client.

### Duplicated with #2720: Secure connection with AWS S3 endpoint
- Added `ObjectStoreConfig.Secure` to the config.
- added `MINIO_SERVICE_SECURE` env flag to set region to use for minio client.

PS: @descampsk, hope u don't mind. Because all these 3 feature requests are so closely related, and I just happens to see ur PR and the CI errors. U need to install bazel in ur local machine to do the build and run the test. U can't see the detailed error logs from the CI - u need to run bazel locally and see the logs. I am not sure what is ur error thou - ur PR looks fine to me.

**Comment by descampsk** [comment]
@eterna2 
Hello,
No worries for me, I'm glad you added that feature I was trying to push.
If it also allows to merge it faster, that's just a bonus.
I'm leaving my PR open for now, and I'll close it once yours is merged.

**Comment by descampsk** [comment]
Will it work with minioHost = s3.amazonaws.com, minioPort=443 et minioSecure=true ?

**Comment by eterna2** [comment]
Yeah. It should work. Cuz under the hood for minio-go it just parses the final endpoint to `https://s3.awsamazon.com:443`

**Comment by eterna2** [comment]
### Changes (Updated 16 Jan 2020)

- made `ObjectStore` folder path configurable from both config.json and env
- updated `minioclient` to use chained provider credentials: API key -> minio env flags -> aws env flags -> aws IAM
- `minioclient` `secure` and `region` flags can now also be set via `config.json` and env var.

```
	minioServiceHost       = "MINIO_SERVICE_SERVICE_HOST"
	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
	minioServiceRegion     = "MINIO_SERVICE_REGION"
	minioServiceSecure     = "MINIO_SERVICE_SECURE"
	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
	pipelineFolder         = "MINIO_PIPELINE_FOLDER"
```

**Comment by eterna2** [comment]
This PR should also fixes https://github.com/e2fyi/kubeflow-aws/issues/1

**Comment by eterna2** [comment]
TODO need to check `ResourceManager.ReadArtifact` calls to `r.objectStore.GetFile(artifactPath)` as the PR introduces a default base folder to the objectStore API.

File: backend/src/apiserver/client/minio.go [hunk 2]
```diff
@@ -16,31 +16,61 @@ package client
...
 
+func getEndpoint(host, port string) string {
+	if port == "" {
+		return host
+	}
+	return fmt.Sprintf("%s:%s", host, port)
+}
+
+// createCredentialProvidersChain creates a chained providers credential for a minio client
+func createCredentialProvidersChain(endpoint, accessKey, secretKey string) *credentials.Credentials {
```
Comment (Jeffwan) [inline backend/src/apiserver/client/minio.go]: seems endpoint is not being used. could we clean it up?

File: backend/src/apiserver/client_manager.go [hunk 3]
```diff
@@ -37,6 +37,10 @@ import (
 const (
 	minioServiceHost       = "MINIO_SERVICE_SERVICE_HOST"
 	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
+	minioServiceRegion     = "MINIO_SERVICE_REGION"
+	minioServiceSecure     = "MINIO_SERVICE_SECURE"
+	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
```
Comment (Jeffwan) [inline backend/src/apiserver/client_manager.go]: Any reason you add this environment?  Does config override work for you?  If so, it's unnecessary to add a new env for same field

File: backend/src/apiserver/client_manager.go [hunk 4]
```diff
@@ -309,29 +313,34 @@ func initMinioClient(initConnectionTimeout time.Duration) storage.ObjectStoreInt
...
-	secretKey := common.GetStringConfig("ObjectStoreConfig.SecretAccessKey")
-	bucketName := common.GetStringConfig("ObjectStoreConfig.BucketName")
+	minioServiceRegion := common.GetStringConfigWithDefault(
+		"ObjectStoreConfig.Region", os.Getenv(minioServiceRegion))
+	minioServiceSecure := common.GetBoolConfigWithDefault(
+		"ObjectStoreConfig.Secure", common.GetBoolFromStringWithDefault(os.Getenv(minioServiceSecure), false))
+	accessKey := common.GetStringConfigWithDefault("ObjectStoreConfig.AccessKey", "")
+	secretKey := common.GetStringConfigWithDefault("ObjectStoreConfig.SecretAccessKey", "")
+	bucketName := common.GetStringConfigWithDefault("ObjectStoreConfig.BucketName", os.Getenv(pipelineBucketName))
+	pipelineFolder := common.GetStringConfigWithDefault("ObjectStoreConfig.PipelineFolder", os.Getenv(pipelineFolder))
```
Comment (Jeffwan) [inline backend/src/apiserver/client_manager.go]: I am thinking maybe we can change `pipelineFolder`  to `pipelinePath` ? object key could be multi-folder like `folderA/folderB/object_name`

**Comment by Jeffwan** [comment]
@eterna2 Trying to get better understand of your case. 

> Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.

Do you mean you have multiple kubeflow clusters for one env and you want to share same bucket and path for pipeline data? 

> I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths

I think it's expected and suggested, different bucket probably want to have different policy. Do you mean you want to use same bucket and have different folder for different env? like
```
bucket/dev-pipeline
bucket/stg-pipeline
bucket/prd-pipeline
```

> and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).

Any reason you want to put pipeline with artifacts folder together?

File: backend/src/apiserver/client_manager.go (hunk 3 above)
Comment (eterna2) [reply backend/src/apiserver/client_manager.go]: true. this is more for consistency.  Personally, I feel we shld remove all the `MINIO_*` and stick with what `viper` provides us.

File: backend/src/apiserver/client_manager.go (hunk 4 above)
Comment (eterna2) [reply backend/src/apiserver/client_manager.go]: ok

**Comment by eterna2** [comment]
> @eterna2 Trying to get better understand of your case.
> 
> > Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.
> 
> Do you mean you have multiple kubeflow clusters for one env and you want to share same bucket and path for pipeline data?

I am running multiple kubeflow pipelines in 2 clusters. In different namespace for different teams for dev and stg/prd - i.e. each team will have 2 kubeflow pipelines, 1 dev and 1 stg/prd. And we have several teams.

So currently the `prefix` to save the pipeline templates are hardcoded. I have 1 s3 bucket for each cluster. And within a cluster, different namespace saves to a different prefix (by team). So if the pipeline path is not configurable, I will need to give each team 2 s3 buckets (dev and stg/prd). 

tldr. same bucket for pipelines in same cluster but different path for each pipeline in different namespace.

> 
> > I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths
> 
> I think it's expected and suggested, different bucket probably want to have different policy. Do you mean you want to use same bucket and have different folder for different env? like
> 
> ```
> bucket/dev-pipeline
> bucket/stg-pipeline
> bucket/prd-pipeline
> ```

yup. But not dev/stg/prd, rather by teamA/teamB/teamC. I use diff buckets for dev/stg/prd.

> 
> > and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).
> 
> Any reason you want to put pipeline with artifacts folder together?

cuz they are tightly coupled. 

I put everything related to a specific kf pipeline deployment in the same folder. i.e. the s3 looks something like:

```
- dev-bucket
  - pipelines
    - namespace A
      - artifacts/**/*/main.log
      - pipelines/**
    - namespace B
      - artifacts/**/*/main.log
      - pipelines/**

```

File: backend/src/apiserver/client_manager.go (hunk 3 above)
Comment (eterna2) [reply backend/src/apiserver/client_manager.go]: The only reason i prefer `env` over `config` is just it takes less effort than attaching a configmap.

**Comment by eterna2** [comment]
ok. I have rebase and sync with the current master. Removed `getEndpoint` cuz I saw the new master has `joinHostPort` which is the same thing.

Renamed `pipelineFolder` to `pipelinePath`. Updated all the unit tests again.

I have tested this on my own cluster and it works fine. 

Pending the decision on the extra env flags. Otherwise shld be good to go.

**Comment by Jeffwan** [comment]
Thanks!

**Comment by descampsk** [comment]
Thanks it will be very helpful !

File: backend/src/apiserver/client/minio.go (hunk 2 above)
Comment (discordianfish) [reply backend/src/apiserver/client/minio.go]: Any reason this wasn't addressed? Given it's a new contribution I wouldn't expect to find such cruft.

File: backend/src/apiserver/client_manager.go [hunk 5]
```diff
@@ -317,29 +321,34 @@ func initMinioClient(initConnectionTimeout time.Duration) storage.ObjectStoreInt
...
+	createMinioBucket(minioClient, bucketName, minioServiceRegion)
 
-	return storage.NewMinioObjectStore(&storage.MinioClient{Client: minioClient}, bucketName, disableMultipart)
+	return storage.NewMinioObjectStore(&storage.MinioClient{Client: minioClient}, bucketName, pipelinePath, disableMultipart)
 }
 
-func createMinioBucket(minioClient *minio.Client, bucketName string) {
+func createMinioBucket(minioClient *minio.Client, bucketName, region string) {
+	// Check to see if we already own this bucket.
+	exists, err := minioClient.BucketExists(bucketName)
```
Comment (discordianfish) [inline backend/src/apiserver/client_manager.go]: Error isn't checked

**Comment by eterna2** [comment]
@discordianfish

The initial PR was opened many months ago before the `joinHostPort` func was merged into master.

There was a large desync from master when the review was done, hence it seems to be "unused". But it was resolved - i.e. removed, after I rebased with master - as commented after the rebase.

U r probably right for the No error checked for bucket exist cuz I am lazy. I am catching the error only after I try creating the bucket.

Mostly because exist can only be true if there is no error. If it does not exist, or the check fails for any reason, I am going to just try creating the bucket to see if it works. If it fails, then I raise an error.
